### DISCUSSION TITLE
Feature: automatic subdomains for apps

### DISCUSSION THREAD

**Discussion by sio** [IceWhaleTech/CasaOS#176] [Ideas]
#### Is your idea related to a problem? Please describe:

CasaOS opens apps on the same domain with random ports appended (e.g. `:8096` for Jellyfin and `:2342` for PhotoPrism on https://demo.casaos.io). Using subdomains would make application URLs more memorable and human readable.

#### Describe the solution you'd like:

- Generate subdomain name based on app name ("Jellyfin" -> `jellyfin.demo.casaos.io`, "PhotoPrism" -> `photoprism.demo.casaos.io`)
- (Optional) Handle subdomain collisions:
  - Prompt user for another name
  - Or just append a sequential index (`jellyfin1.demo.casaos.io`, `jellyfin2.demo.casaos.io`) for collisions
- Provide a reverse proxy to route HTTP requests from subdomains to the respective containers

#### Describe alternatives you've considered:

- Another user suggested [managing subdomains manually](https://github.com/IceWhaleTech/CasaOS/issues/134). I agree with comments there that manual management would require too much technical expertise and would be unnecessary bothersome.

**Comment by Lauren-ED209** [comment]
That will be elegant, as it should be.

**Comment by AuthorShin** [comment]
>     * Another user suggested [managing subdomains manually](https://github.com/IceWhaleTech/CasaOS/issues/134). I agree with comments there that manual management would require too much technical expertise and would be unnecessary bothersome.

Yeah, but some might prefer the manual way for reasons like extra security (i.e,, HSTS) or other futures (i.e,, Web sockets support - Cache) which might not be possible to integrate with casaOS, that's why I suggested adding a button in casaOS setting to toggle between basic mode (having less future requires less manual work and tinkering) pro/advance mode (having access to more advance futures that requires more tinkering and manual configuration). (I might add this as an individual feature request).
